include/boost/variant/detail/apply_visitor_unary.hpp

```diff
@@ -121,39 +121,27 @@ struct result_multideduce1 {
         begin_it, boost::mpl::int_<boost::mpl::size<types>::type::value - 1>
     >::type                                         last_it;
 
-    // For metaprogramming purposes ONLY! Do not use this method (and class) at runtime!
-    static Visitor& vis() BOOST_NOEXCEPT {
-        // Functions that work with lambdas must be defined in same translation unit.
-        // Because of that, we can not use `boost::decval<Visitor&>()` here.
-        Visitor&(*f)() = 0; // pointer to function
-        return f();
-    }
-
-    static decltype(auto) deduce_impl(last_it, unsigned /*helper*/) {
-        typedef typename boost::mpl::deref<last_it>::type value_t;
-        return vis()( boost::declval< value_t& >() );
-    }
-
-    template <class It>
-    static decltype(auto) deduce_impl(It, unsigned helper) {
+    template <class It, class Dummy = void> // avoid explicit specialization in class scope
+    struct deduce_impl {
         typedef typename boost::mpl::next<It>::type next_t;
         typedef typename boost::mpl::deref<It>::type value_t;
-        if (helper == boost::mpl::distance<begin_it, It>::type::value) {
-            return deduce_impl(next_t(), ++helper);
-        }
+        typedef decltype(true ? boost::declval< Visitor& >()( boost::declval< value_t& >() )
```

Comment:
I interpreted the original code as trying to trigger return type deduction to make sure the visitor always returned __the same__ type for each type in the variant. Is this correct?

If so, this patch will change how deduction is done to figure a common type between the return types of the visitor. It should be possible to change that if that is not desirable.